File: 2024/Day_5/Solution.go
package main

import (
    "fmt"
    "os"
    "path/filepath"
    "strings"
    "strconv"
)

type Rule struct {
    before, after int
}

func contains(slice []int, val int) bool {
    for _, item := range slice {
        if item == val {
            return true
        }
    }
    return false
}

func indexOf(slice []int, val int) int {
    for i, item := range slice {
        if item == val {
            return i
        }
    }
    return -1
}

func findSumOfMedians(group [][]int) int {
    sum := 0
    for _, update := range group {
        index := len(update) / 2
        sum += update[index]
    }
    return sum
}

func copySlice(slice []int) []int {
    copy := make([]int, len(slice))
    for i, v := range slice {
        copy[i] = v
    }
    return copy
}

func partOne(updates [][]int, rules []Rule) (int, [][]int) {
    toInclude := [][]int{}
    broken := [][]int{}

    for _, update := range updates {
        count := len(rules)
        isBroken := false

        for _, rule := range rules {
            if contains(update, rule.before) && contains(update, rule.after) {
                firstIndex := indexOf(update, rule.before)
                secondIndex := indexOf(update, rule.after)
                
                if firstIndex < secondIndex {
                    count--
                } else {
                    broken = append(broken, update)
                    isBroken = true
                    break
                }
            } else {
                count--
            }
        }

        if count == 0 && !isBroken {
            toInclude = append(toInclude, update)
        }
    }

    answer := findSumOfMedians(toInclude)
    return answer, broken
}

func partTwo(unsorted [][]int, rules []Rule) int {
    fixed := [][]int{}

    for _, update := range unsorted {
        copy := copySlice(update)
        changesMade := true

        for changesMade {
            changesMade = false
            for _, rule := range rules {
                if contains(copy, rule.before) && contains(copy, rule.after) {
                    firstIndex := indexOf(copy, rule.before)
                    secondIndex := indexOf(copy, rule.after)
                    if firstIndex > secondIndex {
                        copy[firstIndex], copy[secondIndex] = copy[secondIndex], copy[firstIndex]
                        changesMade = true
                    }
                }
            }
        }
        fixed = append(fixed, copy)
    }

    return findSumOfMedians(fixed)
}

func main() {
    cwd, err := os.Getwd()
    if err != nil {
        fmt.Printf("Error getting current directory: %v\n", err)
        return
    }

    content, err := os.ReadFile(filepath.Join(cwd, "2024", "Day_5", "data.txt"))
    if err != nil {
        fmt.Printf("Error reading file: %v\n", err)
        return
    }

    lines := strings.Split(string(content), "\n")
    
    var rules []Rule
    var updates [][]int

    for _, line := range lines {
        if strings.Contains(line, "|") {
            parts := strings.Split(line, "|")
            if len(parts) == 2 {
                before, err1 := strconv.Atoi(strings.TrimSpace(parts[0]))
                after, err2 := strconv.Atoi(strings.TrimSpace(parts[1]))
                if err1 == nil && err2 == nil {
                    rules = append(rules, Rule{before, after})
                }
            }
        } else if strings.Contains(line, ",") {
            parts := strings.Split(line, ",")
            var nums []int
            valid := true
            for _, part := range parts {
                num, err := strconv.Atoi(strings.TrimSpace(part))
                if err != nil {
                    valid = false
                    break
                }
                nums = append(nums, num)
            }
            if valid && len(nums) > 0 {
                updates = append(updates, nums)
            }
        }
    }

    part1, unsorted := partOne(updates, rules)
    fmt.Println("Part 1:", part1)
    
    part2 := partTwo(unsorted, rules)
    fmt.Println("Part 2:", part2)
}

File: 2024/Day_5/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;

class Rule {
    int before;
    int after;

    Rule(int before, int after) {
        this.before = before;
        this.after = after;
    }
}

public class Solution {
    private List<Rule> rules;
    private List<List<Integer>> updates;

    public Solution() {
        rules = new ArrayList<>();
        updates = new ArrayList<>();
    }

    private void readFile() throws IOException {
        Path currentPath = Paths.get(System.getProperty("user.dir"),  "data.txt");
        List<String> lines = Files.readAllLines(currentPath);

        for (String line : lines) {
            if (line.contains("|")) {
                String[] parts = line.split("\\|");
                int before = Integer.parseInt(parts[0].trim());
                int after = Integer.parseInt(parts[1].trim());
                rules.add(new Rule(before, after));
            } else if (line.contains(",")) {
                List<Integer> numbers = Arrays.stream(line.split(","))
                    .map(String::trim)
                    .map(Integer::parseInt)
                    .collect(Collectors.toList());
                updates.add(numbers);
            }
        }
    }

    private int findSumOfMedians(List<List<Integer>> group) {
        int sum = 0;
        for (List<Integer> update : group) {
            int index = update.size() / 2;
            sum += update.get(index);
        }
        return sum;
    }

    public Result partOne() {
        List<List<Integer>> toInclude = new ArrayList<>();
        List<List<Integer>> broken = new ArrayList<>();

        for (List<Integer> update : updates) {
            int count = rules.size();
            boolean isBroken = false;

            for (Rule rule : rules) {
                if (update.contains(rule.before) && update.contains(rule.after)) {
                    int firstIndex = update.indexOf(rule.before);
                    int secondIndex = update.indexOf(rule.after);

                    if (firstIndex < secondIndex) {
                        count--;
                    } else {
                        broken.add(new ArrayList<>(update));
                        isBroken = true;
                        break;
                    }
                } else {
                    count--;
                }
            }

            if (count == 0 && !isBroken) {
                toInclude.add(new ArrayList<>(update));
            }
        }

        int answer = findSumOfMedians(toInclude);
        return new Result(answer, broken);
    }

    public int partTwo(List<List<Integer>> unsorted) {
        List<List<Integer>> fixed = new ArrayList<>();

        for (List<Integer> update : unsorted) {
            List<Integer> copy = new ArrayList<>(update);
            boolean changesMade = true;

            while (changesMade) {
                changesMade = false;
                for (Rule rule : rules) {
                    if (copy.contains(rule.before) && copy.contains(rule.after)) {
                        int firstIndex = copy.indexOf(rule.before);
                        int secondIndex = copy.indexOf(rule.after);
                        if (firstIndex > secondIndex) {
                            Collections.swap(copy, firstIndex, secondIndex);
                            changesMade = true;
                        }
                    }
                }
            }
            fixed.add(copy);
        }

        return findSumOfMedians(fixed);
    }

    private static class Result {
        int answer;
        List<List<Integer>> broken;

        Result(int answer, List<List<Integer>> broken) {
            this.answer = answer;
            this.broken = broken;
        }
    }

    public static void main(String[] args) {
        try {
            Solution solution = new Solution();
            solution.readFile();

            Result result = solution.partOne();
            System.out.println("Part 1: " + result.answer);

            int part2Answer = solution.partTwo(result.broken);
            System.out.println("Part 2: " + part2Answer);

        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}